core/application/impl/app_state_manager_impl.cpp

```diff
@@ -0,0 +1,169 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "application/impl/app_state_manager_impl.hpp"
+
+#include <csignal>
+#include <functional>
+
+namespace kagome {
+
+  std::weak_ptr<AppStateManager> AppStateManagerImpl::wp_to_myself;
+
+  AppStateManagerImpl::AppStateManagerImpl()
+      : logger_(common::createLogger("Application")) {
+    struct sigaction act {};
+    memset(&act, 0, sizeof(act));
+    act.sa_handler = shuttingDownSignalsHandler;  // NOLINT
+    sigset_t set;
+    sigemptyset(&set);
+    sigaddset(&set, SIGINT);
+    sigaddset(&set, SIGTERM);
+    sigaddset(&set, SIGQUIT);
+    act.sa_mask = set;
+    sigaction(SIGINT, &act, nullptr);
+    sigaction(SIGTERM, &act, nullptr);
+    sigaction(SIGQUIT, &act, nullptr);
+    sigprocmask(SIG_UNBLOCK, &act.sa_mask, nullptr);
+  }
+
+  AppStateManagerImpl::~AppStateManagerImpl() {
+    struct sigaction act {};
+    memset(&act, 0, sizeof(act));
+    act.sa_handler = SIG_DFL;  // NOLINT
+    sigset_t set;
+    sigemptyset(&set);
+    sigaddset(&set, SIGINT);
+    sigaddset(&set, SIGTERM);
+    sigaddset(&set, SIGQUIT);
+    act.sa_mask = set;
+    sigaction(SIGINT, &act, nullptr);
+    sigaction(SIGTERM, &act, nullptr);
+    sigaction(SIGQUIT, &act, nullptr);
+  }
+
+  void AppStateManagerImpl::reset() {
+    std::lock_guard lg(mutex_);
+    while (!prepare_.empty()) prepare_.pop();
+    while (!launch_.empty()) launch_.pop();
+    while (!shutdown_.empty()) shutdown_.pop();
+    state_ = State::Init;
+    shutdown_requested_ = false;
+  }
+
+  void AppStateManagerImpl::atPrepare(Callback &&cb) {
+    std::lock_guard lg(mutex_);
+    if (state_ > State::Init) {
+      throw AppStateException("adding callback for stage 'prepare'");
+    }
+    prepare_.emplace(std::move(cb));
+  }
+
+  void AppStateManagerImpl::atLaunch(Callback &&cb) {
+    std::lock_guard lg(mutex_);
+    if (state_ > State::ReadyToStart) {
+      throw AppStateException("adding callback for stage 'launch'");
+    }
+    launch_.emplace(std::move(cb));
+  }
+
+  void AppStateManagerImpl::atShutdown(Callback &&cb) {
+    std::lock_guard lg(mutex_);
+    if (state_ > State::Works) {
+      throw AppStateException("adding callback for stage 'shutdown'");
+    }
+    shutdown_.emplace(std::move(cb));
+  }
+
+  void AppStateManagerImpl::doPrepare() {
+    std::lock_guard lg(mutex_);
+    if (state_ != State::Init) {
+      throw AppStateException("running stage 'prepare'");
+    }
+    state_ = State::Prepare;
+
+    while (!prepare_.empty()) {
+      auto &cb = prepare_.front();
+      if (state_ == State::Prepare) {
+        cb();
+      }
+      prepare_.pop();
+    }
+
+    if (state_ == State::Prepare) {
```

Comment:
you mean that state can be changed during callbacks execution?

Reply:
Yep. If some stage would be failed, entity must do shutdown().